File: frontend/src/lib/layout-taxonomy.ts
import layoutMetadataJson from "@/generated/layout-metadata.json";

type SharedLayoutMetadata = typeof layoutMetadataJson;
type SubGroupsByGroup = SharedLayoutMetadata["subGroupsByGroup"];
type VariantsBySubGroup = SharedLayoutMetadata["variantsBySubGroup"];
type DesignTraitAxes = SharedLayoutMetadata["designTraitAxes"];
type LayoutRecord = Record<
  string,
  {
    group: LayoutGroup;
    subGroup: LayoutSubGroup;
    variantId: LayoutVariantId;
    isVariantDefault?: boolean;
    notes: LayoutTemplateNotes;
  }
>;
type SubGroupRecord = Record<
  LayoutGroup,
  Record<string, { label: string; description: string }>
>;
type VariantRecord = Record<
  LayoutGroup,
  Record<
    string,
    Record<
      string,
      {
        label: string;
        description: string;
        usage: string[];
        usageBias: string;
        designTraits?: LayoutDesignTraits;
      }
    >
  >
>;
type DesignTraitAxisRecord = Record<
  LayoutDesignTraitAxis,
  Record<string, { label: string; description: string }>
>;

export type LayoutGroup = keyof SharedLayoutMetadata["groupLabels"];
export type LayoutSubGroup = {
  [G in keyof SubGroupsByGroup]: keyof SubGroupsByGroup[G];
}[keyof SubGroupsByGroup];
export type LayoutVariantId = {
  [G in keyof VariantsBySubGroup]: {
    [S in keyof VariantsBySubGroup[G]]: keyof VariantsBySubGroup[G][S];
  }[keyof VariantsBySubGroup[G]];
}[keyof VariantsBySubGroup];
export type LayoutDesignTraitAxis = keyof DesignTraitAxes;
export type LayoutVariantTone = keyof DesignTraitAxes["tone"];
export type LayoutVariantStyle = keyof DesignTraitAxes["style"];
export type LayoutVariantDensity = keyof DesignTraitAxes["density"];

export type LayoutDesignTraits = {
  tone?: LayoutVariantTone;
  style?: LayoutVariantStyle;
  density?: LayoutVariantDensity;
};

export type LayoutVariantDefinition = {
  id: LayoutVariantId;
  label: string;
  description: string;
  usage: string[];
  usageBias: string;
  designTraits: LayoutDesignTraits;
};

export type LayoutTemplateNotes = {
  purpose: string;
  structure_signal: string;
  design_signal: string;
  use_when: string;
  avoid_when: string;
  usage_bias: string;
};

export type LayoutTaxonomyEntry = {
  group: LayoutGroup;
  subGroup: LayoutSubGroup;
  variantId: LayoutVariantId;
  isVariantDefault: boolean;
};

const layoutMetadata = layoutMetadataJson as SharedLayoutMetadata;
const layouts = layoutMetadata.layouts as LayoutRecord;
const subGroupsByGroup = layoutMetadata.subGroupsByGroup as SubGroupRecord;
const variantsBySubGroup = layoutMetadata.variantsBySubGroup as VariantRecord;
const designTraitAxes = layoutMetadata.designTraitAxes as DesignTraitAxisRecord;

export const LAYOUT_GROUP_ORDER = [...layoutMetadata.groupOrder] as LayoutGroup[];

export function getLayoutTaxonomy(layoutId: string): LayoutTaxonomyEntry | null {
  const metadata = layouts[layoutId];

  if (!metadata) {
    return null;
  }

  return {
    group: metadata.group,
    subGroup: metadata.subGroup as LayoutSubGroup,
    variantId: metadata.variantId,
    isVariantDefault: Boolean(metadata.isVariantDefault),
  };
}

export function getLayoutNotes(layoutId: string): LayoutTemplateNotes | null {
  return layouts[layoutId]?.notes ?? null;
}

export function getLayoutVariantId(layoutId: string): LayoutVariantId | null {
  return layouts[layoutId]?.variantId ?? null;
}

export function getLayoutGroupLabel(group: LayoutGroup): string {
  return layoutMetadata.groupLabels[group];
}

export function getLayoutGroupDescription(group: LayoutGroup): string {
  return layoutMetadata.groupDescriptions[group];
}

export function getLayoutSubGroupsForGroup(group: LayoutGroup): LayoutSubGroup[] {
  return Object.keys(subGroupsByGroup[group] ?? {}) as LayoutSubGroup[];
}

export function getLayoutSubGroupLabel(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
): string {
  return subGroupsByGroup[group]?.[subGroup]?.label ?? subGroup;
}

export function getLayoutSubGroupDescription(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
): string {
  return subGroupsByGroup[group]?.[subGroup]?.description ?? "";
}

export function getVariantsForSubGroup(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
): LayoutVariantId[] {
  return Object.keys(
    variantsBySubGroup[group]?.[subGroup] ?? {},
  ) as LayoutVariantId[];
}

export function getVariantDefinition(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): LayoutVariantDefinition | null {
  const definition = variantsBySubGroup[group]?.[subGroup]?.[variantId];
  if (!definition) {
    return null;
  }
  return {
    id: variantId,
    label: definition.label,
    description: definition.description,
    usage: definition.usage,
    usageBias: definition.usageBias,
    designTraits: definition.designTraits ?? {},
  };
}

export function getVariantLabel(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): string {
  return getVariantDefinition(group, subGroup, variantId)?.label ?? variantId;
}

export function getVariantDescription(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): string {
  return getVariantDefinition(group, subGroup, variantId)?.description ?? "";
}

export function getVariantUsage(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): string[] {
  return getVariantDefinition(group, subGroup, variantId)?.usage ?? [];
}

export function getVariantUsageBias(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): string {
  return getVariantDefinition(group, subGroup, variantId)?.usageBias ?? "";
}

export function getVariantDesignTraits(
  group: LayoutGroup,
  subGroup: LayoutSubGroup,
  variantId: LayoutVariantId,
): LayoutDesignTraits {
  return getVariantDefinition(group, subGroup, variantId)?.designTraits ?? {};
}

export function getLayoutDesignTraitLabel(
  axis: LayoutDesignTraitAxis,
  value: string,
): string {
  return designTraitAxes[axis]?.[value]?.label ?? value;
}

export function getLayoutDesignTraitDescription(
  axis: LayoutDesignTraitAxis,
  value: string,
): string {
  return designTraitAxes[axis]?.[value]?.description ?? "";
}

File: backend/app/services/layouts/layout_taxonomy.py
"""Formal three-layer layout taxonomy helpers backed by shared metadata."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, cast

from app.services.layouts.layout_metadata import load_layout_metadata

LayoutGroup = Literal[
    "cover",
    "agenda",
    "section-divider",
    "narrative",
    "evidence",
    "comparison",
    "process",
    "highlight",
    "closing",
]

LayoutDesignTraitAxis = Literal["tone", "style", "density"]

_SHARED_METADATA = load_layout_metadata()


@dataclass(frozen=True)
class LayoutDesignTraits:
    tone: str = ""
    style: str = ""
    density: str = ""


@dataclass(frozen=True)
class LayoutVariantDefinition:
    id: str
    label: str
    description: str
    usage: tuple[str, ...]
    usage_bias: str
    design_traits: LayoutDesignTraits


@dataclass(frozen=True)
class LayoutTemplateNotes:
    purpose: str
    structure_signal: str
    design_signal: str
    use_when: str
    avoid_when: str
    usage_bias: str


@dataclass(frozen=True)
class LayoutTaxonomyEntry:
    group: LayoutGroup
    sub_group: str
    variant_id: str
    is_variant_default: bool


GROUP_ORDER: tuple[LayoutGroup, ...] = tuple(
    cast(LayoutGroup, group) for group in _SHARED_METADATA["groupOrder"]
)


def get_layout_taxonomy(layout_id: str) -> LayoutTaxonomyEntry | None:
    metadata = _SHARED_METADATA["layouts"].get(layout_id)
    if not isinstance(metadata, dict):
        return None

    return LayoutTaxonomyEntry(
        group=cast(LayoutGroup, metadata.get("group", "narrative")),
        sub_group=str(metadata.get("subGroup", "default")),
        variant_id=str(metadata.get("variantId", "")),
        is_variant_default=bool(metadata.get("isVariantDefault", False)),
    )


def get_layout_notes(layout_id: str) -> LayoutTemplateNotes | None:
    metadata = _SHARED_METADATA["layouts"].get(layout_id)
    if not isinstance(metadata, dict):
        return None

    notes = metadata.get("notes", {})
    if not isinstance(notes, dict):
        return None

    return LayoutTemplateNotes(
        purpose=str(notes.get("purpose", "")),
        structure_signal=str(notes.get("structure_signal", "")),
        design_signal=str(notes.get("design_signal", "")),
        use_when=str(notes.get("use_when", "")),
        avoid_when=str(notes.get("avoid_when", "")),
        usage_bias=str(notes.get("usage_bias", "")),
    )


def get_group_order() -> tuple[LayoutGroup, ...]:
    return GROUP_ORDER


def get_layout_group_label(group: LayoutGroup) -> str:
    return str(_SHARED_METADATA["groupLabels"][group])


def get_layout_group_description(group: LayoutGroup) -> str:
    return str(_SHARED_METADATA["groupDescriptions"][group])


def get_sub_groups_for_group(group: LayoutGroup) -> tuple[str, ...]:
    return tuple(_SHARED_METADATA["subGroupsByGroup"].get(group, {}).keys())


def get_sub_group_label(group: LayoutGroup, sub_group: str) -> str:
    return str(
        _SHARED_METADATA["subGroupsByGroup"].get(group, {}).get(sub_group, {}).get("label", sub_group)
    )


def get_sub_group_description(group: LayoutGroup, sub_group: str) -> str:
    return str(
        _SHARED_METADATA["subGroupsByGroup"]
        .get(group, {})
        .get(sub_group, {})
        .get("description", "")
    )


def get_layout_variant_definition(
    group: LayoutGroup,
    sub_group: str,
    variant_id: str,
) -> LayoutVariantDefinition | None:
    definition = (
        _SHARED_METADATA["variantsBySubGroup"]
        .get(group, {})
        .get(sub_group, {})
        .get(variant_id)
    )
    if not isinstance(definition, dict):
        return None

    design_traits = definition.get("designTraits", {})
    return LayoutVariantDefinition(
        id=variant_id,
        label=str(definition.get("label", variant_id)),
        description=str(definition.get("description", "")),
        usage=tuple(str(tag) for tag in definition.get("usage", [])),
        usage_bias=str(definition.get("usageBias", "")),
        design_traits=LayoutDesignTraits(
            tone=str(design_traits.get("tone", "")),
            style=str(design_traits.get("style", "")),
            density=str(design_traits.get("density", "")),
        ),
    )


def get_variant_ids_for_sub_group(group: LayoutGroup, sub_group: str) -> tuple[str, ...]:
    return tuple(
        _SHARED_METADATA["variantsBySubGroup"].get(group, {}).get(sub_group, {}).keys()
    )


def get_design_trait_label(axis: LayoutDesignTraitAxis, value: str) -> str:
    return str(_SHARED_METADATA["designTraitAxes"].get(axis, {}).get(value, {}).get("label", value))


def get_design_trait_description(axis: LayoutDesignTraitAxis, value: str) -> str:
    return str(
        _SHARED_METADATA["designTraitAxes"].get(axis, {}).get(value, {}).get("description", "")
    )
